### ISSUE TITLE
Removing a node fails

### ISSUE THREAD

**Issue by viveleroi** [Matt-Esch/virtual-dom#333]
I'm building an array of `li` elements for an `ol`. 

If any future render has fewer list items, the diff/patching process breaks. Rather than removing dom elements for the nodes I've actually removed, it removes an equal number of nodes from the end.

**Comment by Raynos** [comment]
This is working as intended. The diff process removes nodes from the end and mutates nodes in the middle.

If you want the nodes to be moved or generally want a minimal diff you should use the keys feature. This should have the behavior you are expecting.

**Comment by viveleroi** [comment]
Makes sense. To clarify, you mean the key feature in `h`? So if I set a unique key via `h('div', { key: someKey })`, vdom will be able to recognize when nodes are moved/removed and re-render properly?
